Write Library changes to S/1.0 rewrite Branch

2018denjasa opened the pull request:
This PR includes a few minor changes to the write library which allows rewritten DICOMs to be identical to the original files. Notable changes include:

- made newElement() and mustNewElement() exported functions so that they can be used to create artificial Datasets for writing
- Added writing of the FileMetaInformationVersion header in the correct spot to avoid "Dataset not in ascending order" error
- Removed an error check that the Values array passes to the writeBytes() function is only of length 1
- Removed a duplicate call to encodeElementHeader() when writing PixelData
- Added "DS" and "CS" VRs to the list of string type headers that should be padded with a space, as opposed to NULL

I'll be testing this solution at scale on thousands of open DICOMs to ensure compatibility with as many files as possible. 

Here is a dcmdump of the DICOM I used for testing. Note that there are two values for FileMetaInformationVersion:
```
# Dicom-File-Format

# Dicom-Meta-Information-Header
# Used TransferSyntax: Little Endian Explicit
(0002,0000) UL 196                                      #   4, 1 FileMetaInformationGroupLength
(0002,0001) OB 00\01                                    #   2, 1 FileMetaInformationVersion
(0002,0002) UI =ComputedRadiographyImageStorage         #  26, 1 MediaStorageSOPClassUID
(0002,0003) UI [1.2.276.0.7230010.3.1.4.8323329.3438.1594319553.23768] #  54, 1 MediaStorageSOPInstanceUID
(0002,0010) UI =LittleEndianImplicit                    #  18, 1 TransferSyntaxUID
(0002,0012) UI [1.2.276.0.7230010.3.0.3.6.2]            #  28, 1 ImplementationClassUID
(0002,0013) SH [OFFIS_DCMTK_362]                        #  16, 1 ImplementationVersionName

# Dicom-Data-Set
# Used TransferSyntax: Little Endian Implicit
(0008,0000) UL 288                                      #   4, 1 GenericGroupLength
(0008,0008) CS [ORIGINAL PRIMARY]                       #  16, 1 ImageType
(0008,0012) DA [1995.07.20]                             #  10, 1 InstanceCreationDate
(0008,0016) UI =ComputedRadiographyImageStorage         #  26, 1 SOPClassUID
(0008,0018) UI [1.2.276.0.7230010.3.1.4.8323329.3438.1594319553.23768] #  54, 1 SOPInstanceUID
(0008,0022) DA [1993.02.24]                             #  10, 1 AcquisitionDate
(0008,0032) TM [17:58:00]                               #   8, 1 AcquisitionTime
(0008,0050) SH [FUJI95714]                              #  10, 1 AccessionNumber
(0008,0060) CS [CR]                                     #   2, 1 Modality
(0008,0070) LO [FUJI PHOTO FILM CO. LTD.]               #  24, 1 Manufacturer
(0008,0080) LO [FUJI PHOTO FILM CO.,]                   #  20, 1 InstitutionName
(0008,1040) LO [!!!!!!!!]                               #   8, 1 InstitutionalDepartmentName
(0008,1090) LO [9000]                                   #   4, 1 ManufacturerModelName
(0010,0000) UL 0                                        #   4, 1 GenericGroupLength
(0018,0000) UL 140                                      #   4, 1 GenericGroupLength
(0018,0015) CS [CHEST]                                  #   6, 1 BodyPartExamined
(0018,1050) DS [0.2]                                    #   4, 1 SpatialResolution
(0018,1260) SH [ST]                                     #   2, 1 PlateType
(0018,1261) LO [5]                                      #   2, 1 PhosphorType
(0018,1400) LO [[CHEST,GENERAL     ]]                   #  20, 1 AcquisitionDeviceProcessingDescription
(0018,1401) LO [00]                                     #   2, 1 AcquisitionDeviceProcessingCode
(0018,1402) CS [PORTRAIT]                               #   8, 1 CassetteOrientation
(0018,1403) CS [35CMX35CM]                              #  10, 1 CassetteSize
(0018,5101) CS [PA]                                     #   2, 1 ViewPosition
(0018,6000) DS [357]                                    #   4, 1 Sensitivity
(0020,0000) UL 134                                      #   4, 1 GenericGroupLength
(0020,000d) UI [1.2.276.0.7230010.3.1.2.8323329.3438.1594319553.23766] #  54, 1 StudyInstanceUID
(0020,000e) UI [1.2.276.0.7230010.3.1.3.8323329.3438.1594319553.23767] #  54, 1 SeriesInstanceUID
(0020,0013) IS [16]                                     #   2, 1 InstanceNumber
(0028,0000) UL 124                                      #   4, 1 GenericGroupLength
(0028,0002) US 1                                        #   2, 1 SamplesPerPixel
(0028,0004) CS [MONOCHROME1]                            #  12, 1 PhotometricInterpretation
(0028,0010) US 440                                      #   2, 1 Rows
(0028,0011) US 440                                      #   2, 1 Columns
(0028,0034) IS [01]                                     #   2, 1 PixelAspectRatio
(0028,0100) US 16                                       #   2, 1 BitsAllocated
(0028,0101) US 10                                       #   2, 1 BitsStored
(0028,0102) US 9                                        #   2, 1 HighBit
(0028,0103) US 0                                        #   2, 1 PixelRepresentation
(0028,1050) DS [550]                                    #   4, 1 WindowCenter
(0028,1051) DS [1024]                                   #   4, 1 WindowWidth
(2010,0000) UL 14                                       #   4, 1 GenericGroupLength
(2010,0100) CS [BLACK]                                  #   6, 1 BorderDensity
(7fe0,0000) UL 387208                                   #   4, 1 GenericGroupLength
(7fe0,0010) OW 0000\00d1\030e\03b2\03ba\03bc\03bc\03bc\03be\03bf\03c0\03bc\03ab... # 387200, 1 PixelData
```

2018denjasa commented:
TODO: Implement float and sequence writing

suyashkumar commented on element.go:
```diff
@@ -126,8 +126,8 @@ func newElement(t tag.Tag, data interface{}) (*Element, error) {
 
-func mustNewElement(t tag.Tag, data interface{}) *Element {
-	elem, err := newElement(t, data)
+func MustNewElement(t tag.Tag, data interface{}) *Element {
```
let's leave this unexported if only used within this package, would rather not clutter the exported namespace of the package. Callers can just use NewElement with the error, or write a small must new element shim if needed on their end, doesn't necessarily need to be part of our API I think. As much as possible I'd like to have a minimal exported API, but let me know if you have some good use cases or thoughts otherwise!

2018denjasa replied:
Sounds good, I only made this exported since the old api utilized the MustNewElement method, but I think just using NewElement and handling the error will be fine.

suyashkumar commented on write.go:
```diff
@@ -96,8 +96,16 @@ func writeFileHeader(w dicomio.Writer, ds *Dataset, metaElems []*Element, opts .
-	err := writeMetaElem(subWriter, tag.MediaStorageSOPClassUID, ds, &tagsUsed, opts...)
+	// XXXHACK TODO: decide how to handle this header, since it needs to be written here for correct ordering
+	_, err := ds.FindElementByTag(tag.FileMetaInformationVersion)
+	if err != ErrorElementNotFound {
```
is this intended? This will only execute this code if the error is something other that Element not found, which could include other errors in the future. I expect if there is any error here in finding the tag, we do not want to write this (and probably want to handle it in a specific way if needed).

Also I don't think this is needed, because writeMetaElem does this existence checking within it already?

Also, this tag is found in metaElems already right? I'm unsure if this function needs to take both metaElems and ds *Dataset, but I can clean this and other things up in a future PR as well.

2018denjasa replied:
You're right, my logic in this block of code wasn't entirely sound. I've changed this code block to just try and write the header (as you mentioned the writeMetaElem does the existence checking) and then I handle the error afterwards. If the error is anything but ErrorElementNotFound then I return the error, otherwise I ignore it move on to the next one

suyashkumar commented on write.go:
```diff
@@ -165,7 +173,12 @@ func writeElement(w dicomio.Writer, elem *Element, opts ...WriteOption) error {
 	}
 
-	err = encodeElementHeader(w, elem.Tag, vr, uint32(len(data.Bytes())))
+	length := uint32(len(data.Bytes()))
+	if elem.ValueLength == tag.VLUndefinedLength {
+		length = elem.ValueLength
+	}
```
nit: may read a little cleaner imo as 

```suggestion
	length := uint32(len(data.Bytes()))
	if elem.ValueLength == tag.VLUndefinedLength {
		length = tag.VLUndefinedLength
	}
```

2018denjasa replied:
Sounds good

suyashkumar commented on write.go:
```diff
@@ -418,10 +428,6 @@ func writePixelData(w dicomio.Writer, t tag.Tag, value Value, vr string, vl uint
 		numValues := len(image.Frames[0].NativeData.Data[0])
 		length := numFrames * numPixels * numValues * image.Frames[0].NativeData.BitsPerSample / 8 // length in bytes
 
-		err := encodeElementHeader(w, t, vr, uint32(length))
-		if err != nil {
-			return err
-		}
```
don't we still need to encode the element header in the else block?

2018denjasa replied:
Based on my testing and understanding so far, I don't think we need to encode the element header in the if or the else blocks since the element header gets written at line 181 in the writeElement function. I found that the first 8 bytes of pixel data was a duplicate element header because of the redundancy

suyashkumar reviewed (commented):
Thanks so much for the PR! Looks generally good to me, but had a couple quick nits and comments for your consideration.

suyashkumar commented:
Also, regarding float support see #120. I can also add float support to the write package in a future PR, ideally after this one!

suyashkumar commented:
#121 wip to add write support to floats (mostly done). 

Am thinking it might be cleaner to later introduce two float value types, one for float32 and one for float64, but that can come in a follow up PR, so long as this unblocks some of the use cases.

2018denjasa commented on read.go:
```diff
@@ -376,8 +371,16 @@ func readBytes(r dicomio.Reader, t tag.Tag, vr string, vl uint32) (Value, error)
+	if !onlySpaces{
+		// String may have '\0' suffix if its length is odd.
+		str = strings.Trim(str, " \000")
+	}
```
I made this change because in one test case the value for "Referring Physician Name" was just two spaces "  " and this was trimmed down to an empty string by the existing code block. @suyashkumar I'd like to hear your thoughts on how to navigate this case

suyashkumar replied:
Probably the best way to handle this is to only trim \000 and not trim the space character. (e.g. just remove the space in front of \000

2018denjasa replied:
I tries this as a fix as well before resorting to my current solution and for some reason the two spaces still got trimmed down to the empty string

suyashkumar replied:
Gotcha. If you happen to know of an open source / openly available DICOM that has this behavior, it might be good for us to include in a test set. If so, if you could open up a separate issue and attach the dicom there, that would be awesome!

2018denjasa commented on read.go:
```diff
@@ -110,12 +111,6 @@ func readPixelData(r dicomio.Reader, t tag.Tag, vr string, vl uint32, d *Dataset
-		_, _, err := readRawItem(r)
-		if err != nil {
-			return nil, err
-		}
```
@suyashkumar I removed this code because some of the files I was testing the write functionality on did not use a basic offset table. By removing this code the offset tables were just read in as frames and then were rewritten byte for byte identical to the original file

2018denjasa replied:
That sounds good. I'll revert the changes in this PR since there is no real impact on the resulting written DICOMs. I think the changes you proposed would be great to have in a follow up PR and I'd be happy to help out with the implementation

2018denjasa commented on read.go:
```diff
@@ -39,6 +40,9 @@ func readVR(r dicomio.Reader, isImplicit bool, t tag.Tag) (string, error) {
 	}
+	if t == tag.Item {  //http://dicom.nema.org/dicom/2013/output/chtml/part05/sect_7.5.html
+		return "na", nil
+	}
```
I found this error during testing and I'm not quite sure what the best fix is. Based on this web page: http://dicom.nema.org/dicom/2013/output/chtml/part05/sect_7.5.html
It says that Sequence Items don't have explicit VRs written and the 4 bytes immediately after the tag is the value length as a UINT32

suyashkumar replied:
Thanks for raising this! Do you have an example DICOM that has an issue?

I think all we need to do is just ensure that elements are read with an implicit transfer syntax if they are an item.

e.g something like

if t == tag.Item {
    force_implicit = true
}

I can send a quick PR to address this, so this one can be mostly focued on the write specific logic--what I'm worried about is why we haven't seen an issue yet in previous testing. (Maybe because we had files encoded in implicit transfer syntax for the sequence testing?)

Either way, if you have a problematic file we can add to the set of test files, that would be great!

suyashkumar replied:
See #122, I __think__ this will solve this, but some test data is essential (that was breaking before)

2018denjasa replied:
Sounds good. I agree it may be better to address this in another PR. I've taken a look at the changes in #122 and made some comments. I'll try and attach a test file that demonstrated this issue. The problem is actually very subtle because no error gets raised, the parsing just ends prematurely after the first sequence and it thinks it has read the entire file.

2018denjasa commented on read.go:
```diff
@@ -65,6 +69,16 @@ func readVL(r dicomio.Reader, isImplicit bool, t tag.Tag, vr string) (uint32, er
+		if vl == 0xffffffff {
+			vl = tag.VLUndefinedLength
+		}
+		return vl, nil
```
This is resolving the same issue as above re: sequence Item encodings

suyashkumar reviewed (commented):
Thanks for the updates!! Left a couple more comments for your review.

2018denjasa commented:
@suyashkumar I've removed some of the changes from this PR (offset table and sequence item implicit VR) so that this PR can be focused on write changes. 

Here's how I tested these changes:
- I rewrote 10k DICOM files using pydicom
- I rewrote the same 10k DICOM files using this library
- In both rewrites I filtered out all Sequences
- I then read both rewritten files using pydicom and compared the loaded datasets

All 10k test files were rewritten with the same dataset except for a few files which differed at the offset table within pixel data. Let me know if there are any other tests you think we should run and/or any more changes we should make before merging. Thanks!

suyashkumar reviewed (approved):
Thanks so much for your work and effort on this! Generally LGTM, and there are some things I will address in follow up PRs as discussed earlier.

suyashkumar commented:
Tagged changes with `v1.0.0-alpha.6` :)

suyashkumar commented:
`v1.0.0-alpha.7` includes this and #122 :)
